File: Aoc_Kotlin/src/main/kotlin/day12/cavePaths/Day12.kt
package day12.cavePaths

import AoC_Lib.*

object Day12: SomeDay(2021,12) {
    override val title = "Passage Pathing"

    private fun bfs(cave: Map<String, List<String>>, start: String, end: String, filterNeighbors: (String, List<String>) -> Boolean): List<List<String>> {
        tailrec fun rec(queue: List<List<String>>, visited: List<List<String>>): List<List<String>> {
            val neighbors: List<List<String>> = queue
                .filterNot { it.first() == end }
                .flatMap { path -> cave[path.first()]!!
                    .filter { neighbor -> filterNeighbors(neighbor, path)}
                .map { neighbor -> listOf(neighbor) + path }}
            val newVisited: List<List<String>> = visited + queue
            val newQueue: List<List<String>> = neighbors - visited.toSet()
            return if (newQueue.isEmpty()) newVisited
            else rec(newQueue, newVisited)
        }
        return rec(listOf(listOf(start)), emptyList())
    }

    override fun first(data: String): Any? {
        val cave = data.toStrs()
            .flatMap { it.split('-')
                .let {a -> listOf(a[0] to a[1],
                    a[1] to a[0]) }}
            .groupBy({ it.first },{ it.second })
        fun fp1(n: String, p: List<String>): Boolean =
            n[0].isUpperCase() || n !in p
        return bfs(cave, "start", "end", ::fp1)
            .count{ it.first() == "end" }
    } // 3369 Time: 60ms

    override fun second(data: String): Any? {
        val cave = data.toStrs()
            .flatMap { it.split('-')
                .let {a -> listOf(a[0] to a[1],
                    a[1] to a[0]) }}
            .groupBy({ it.first },{ it.second })
        fun fp2(n: String, p: List<String>): Boolean =
            when {
                n == "start" -> false
                n.first().isUpperCase() -> true
                n !in p -> true
                else -> p.filter { it.first().isLowerCase() }
                    .groupBy { it }
                    .none { it.value.size == 2 }

            }
        return bfs(cave, "start", "end", ::fp2)
            .count{ it.first() == "end" }
    } // 85883 Time: 642ms
}

fun main() = SomeDay.mainify(Day12)

File: AoC_Scala3/src/day12/day12.scala
package day12

import AoC_Lib.*
import scala.annotation.tailrec

type Path = List[String]

object day12 {
  val m: Seq[(String, String)] = inputStrs(day = 12)
    .map(_.split('-'))
    .collect{ case Array(a1,a2) => a1 -> a2 }

  val cave: Map[String, Seq[String]] = (m ++ m.map(_.swap))
    .groupMap(_._1)(_._2)

  def bfs(start: String, filterNeighbors: (String, List[String]) => Boolean): Seq[Path] = {
    @tailrec def rec(queue: List[Path], visited: List[Path]): List[Path] = {
      val neighbors = queue
        .filterNot(_.head == "end")
        .flatMap(path => cave(path.head)
          .filter(neighbor => filterNeighbors(neighbor, path))
          .map(neighbor => neighbor :: path))
      val newVisited = visited ++ queue
      val newQueue = neighbors diff visited
      if (newQueue.isEmpty) newVisited
      else rec(newQueue, newVisited)
    }
    rec(List(List(start)), List.empty)
  }

  def part1: Int = {
    def part1Filter(n: String, p: List[String]): Boolean =
      n.head.isUpper || !p.contains(n)
    bfs("start", part1Filter).count(_.head == "end")
  }

  def part2: Int = {
    def part2Filter(n: String, p: List[String]): Boolean =
      if n == "start" then false
      else if n.head.isUpper then true
      else if !p.contains(n) then true
      else p.filter(_.head.isLower)
        .groupBy(identity)
        .forall(_._2.sizeIs < 2)
    bfs("start",part2Filter).count(_.head == "end")
  }

  @main
  def PassagePathing(): Unit = {
    Console.println(part1) // 3369
    Console.println(part2) // 85883
  }
}